lib/stdlib/src/peer.erl

```diff
@@ -0,0 +1,846 @@
+
+name_arg(error, error, _) ->
+    []; %% no name, no host - starting node that is not distributed
+name_arg(error, {ok, Host}, LongOrShort) ->
+    %% interesting fallback: host is set, name is not. Do what predecessor did,
+    %%  take the current node name and use it...
+    [Name, _] = string:lexemes(atom_to_list(node()), "@"),
+    name_arg(Name ++ "@" ++ Host, error, LongOrShort);
+name_arg({ok, Name}, Host, LongOrShort) ->
+    name_arg(Name, Host, LongOrShort); %% unpack node name
+name_arg(Name, Host, LongOrShort) when is_atom(Name) ->
+    name_arg(atom_to_list(Name), Host, LongOrShort); %% convert atom to list for command line
+name_arg(Name, Host, {ok, ignored}) ->
+    name_arg(Name, Host, {ok, false}); %% fallback to shortnames when origin is not distributed
+name_arg(Name, Host, error) ->
+    name_arg(Name, Host, {ok, net_kernel:longnames()}); %% no longnames present it start options
+name_arg(Name, {ok, Host}, LongOrShort) ->
+    name_arg(Name ++ "@" ++ Host, error, LongOrShort); %% merge host part
+%% only these are actual name creation clauses
+name_arg(Name, error, {ok, true}) ->
+    ["-name", Name];
+name_arg(Name, error, {ok, false}) ->
+    ["-sname", Name].
+
+command_line(Listen, Options) ->
+    %% Node name/sname
+    NameArg = name_arg(maps:find(name, Options), maps:find(host, Options), maps:find(longnames, Options)),
+    %% additional command line args
+    CmdOpts = maps:get(args, Options, []),
+    % start command
+    StartCmd =
+        case Listen of
+            undefined when map_get(connection, Options) =:= standard_io ->
+                ["-user", atom_to_list(?MODULE)];
+            undefined ->
+                Self = base64:encode_to_string(term_to_binary(self())),
+                ["-detached", "-noinput", "-user", atom_to_list(?MODULE), "-origin", Self];
+            {Ips, Port} ->
+                IpStr = lists:concat(lists:join(",", [inet:ntoa(Ip) || Ip <- Ips])),
+                ["-detached", "-noinput", "-user", atom_to_list(?MODULE), "-origin", IpStr, integer_to_list(Port)]
+        end,
+    %% build command line
+    {Exec, PreArgs} = exec(Options),
+    {Exec, PreArgs ++ NameArg ++ StartCmd ++ CmdOpts}.
+
+exec(#{exec := Prog}) when is_list(Prog) ->
+    {Prog, []};
+exec(#{exec := {Prog, Args}}) when is_list(Prog), is_list(Args) ->
+    {Prog, Args};
+exec(Options) when not is_map_key(exec, Options) ->
+    case init:get_argument(progname) of
+        {ok, [[Prog]]} ->
+            {os:find_executable(Prog), []};
+        _ ->
+            %% if progname is not known, use `erlexec` from the same ERTS version we're currently running
+            %% BINDIR environment variable is already set
+            %% EMU variable it also set
+            Root = code:root_dir(),
+            Erts = filename:join(Root, lists:concat(["erts-", erlang:system_info(version)])),
+            BinDir = filename:join(Erts, "bin"),
+            {filename:join(BinDir, "erlexec"), []}
+    end.
+
```

Comment:
Here is hack to handle our special builds such as address sanitizer, gcov and so on. It has been tested in our daily builds.

```suggestion
            find_executable_and_args(Prog);
        _ ->
            %% if progname is not known, use `erlexec` from the same ERTS version we're currently running
            %% BINDIR environment variable is already set
            %% EMU variable it also set
            %% (NOTE: this will not work in a source code tree)
            Root = code:root_dir(),
            Erts = filename:join(Root, lists:concat(["erts-", erlang:system_info(version)])),
            BinDir = filename:join(Erts, "bin"),
            {filename:join(BinDir, "erlexec"), []}
    end.

find_executable_and_args(Prog) ->
    case os:find_executable(Prog) of
        false ->
            case binary:split(list_to_binary(Prog), <<"/cerl ">>) of
                [Prefix, Arg] ->
                    %% This is a hack to handle the 'cerl' script used
                    %% by the Erlang/OTP test suites. When 'cerl'
                    %% starts the runtime system, it typically sets
                    %% 'progname' to the path of the 'cerl' script,
                    %% followed by an argument. For example:
                    %%
                    %%     /<full_path_to>/cerl -debug
                    %%     /<full_path_to>/cerl -asan
                    %%     /<full_path_to>/cerl -gcov
                    %%
                    %% We should find a better way to handle this, for
                    %% example by passing the emulator type and flavor
                    %% using the '-emu_type' and '-emu_flavor'
                    %% options. However, this is not without
                    %% complications as those option does not really
                    %% work for an installed system. Also, it is
                    %% probably a good idea to stop using 'slave'
                    %% before attempting to do this.

                    {binary_to_list(Prefix) ++ "/cerl", [binary_to_list(Arg)]};
                _ ->
                    %% FIXME: Handle this error in a better way.
                    error(failed_to_find_executable)
            end;
        Executable ->
            {Executable, []}
    end.
```